### PR TITLE
[SPARK-19761][SQL]create InMemoryFileIndex with an empty rootPaths when set PARALLEL_PARTITION_DISCOVERY_THRESHOLD to zero failed

### PR DESCRIPTION
**Pull request by windpiger**
## What changes were proposed in this pull request?

If we create a InMemoryFileIndex with an empty rootPaths when set PARALLEL_PARTITION_DISCOVERY_THRESHOLD to zero, it will throw an  exception:

```
Positive number of slices required
java.lang.IllegalArgumentException: Positive number of slices required
        at org.apache.spark.rdd.ParallelCollectionRDD$.slice(ParallelCollectionRDD.scala:119)
        at org.apache.spark.rdd.ParallelCollectionRDD.getPartitions(ParallelCollectionRDD.scala:97)
        at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
        at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
        at scala.Option.getOrElse(Option.scala:121)
        at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
        at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
        at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
        at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
        at scala.Option.getOrElse(Option.scala:121)
        at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
        at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
        at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
        at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
        at scala.Option.getOrElse(Option.scala:121)
        at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
        at org.apache.spark.SparkContext.runJob(SparkContext.scala:2084)
        at org.apache.spark.rdd.RDD$$anonfun$collect$1.apply(RDD.scala:936)
        at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
        at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
        at org.apache.spark.rdd.RDD.withScope(RDD.scala:362)
        at org.apache.spark.rdd.RDD.collect(RDD.scala:935)
        at org.apache.spark.sql.execution.datasources.PartitioningAwareFileIndex$.org$apache$spark$sql$execution$datasources$PartitioningAwareFileIndex$$bulkListLeafFiles(PartitioningAwareFileIndex.scala:357)
        at org.apache.spark.sql.execution.datasources.PartitioningAwareFileIndex.listLeafFiles(PartitioningAwareFileIndex.scala:256)
        at org.apache.spark.sql.execution.datasources.InMemoryFileIndex.refresh0(InMemoryFileIndex.scala:74)
        at org.apache.spark.sql.execution.datasources.InMemoryFileIndex.<init>(InMemoryFileIndex.scala:50)
        at org.apache.spark.sql.execution.datasources.FileIndexSuite$$anonfun$9$$anonfun$apply$mcV$sp$2.apply$mcV$sp(FileIndexSuite.scala:186)
        at org.apache.spark.sql.test.SQLTestUtils$class.withSQLConf(SQLTestUtils.scala:105)
        at org.apache.spark.sql.execution.datasources.FileIndexSuite.withSQLConf(FileIndexSuite.scala:33)
        at org.apache.spark.sql.execution.datasources.FileIndexSuite$$anonfun$9.apply$mcV$sp(FileIndexSuite.scala:185)
        at org.apache.spark.sql.execution.datasources.FileIndexSuite$$anonfun$9.apply(FileIndexSuite.scala:185)
        at org.apache.spark.sql.execution.datasources.FileIndexSuite$$anonfun$9.apply(FileIndexSuite.scala:185)
        at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
        at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
```

## How was this patch tested?
unit test added

### CODE REVIEW

**Comment by windpiger** [comment]
cc @cloud-fan @gatorsmile

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala [hunk 1]
```diff
@@ -178,6 +179,12 @@ class FileIndexSuite extends SharedSQLContext {
       assert(catalog2.allFiles().nonEmpty)
     }
   }
+
+  test("InMemoryFileIndex with empty rootPaths when PARALLEL_PARTITION_DISCOVERY_THRESHOLD is 0") {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: After this fix, when users setting it to `-1`, we still face the same strange error. We need a complete fix.

Comment (windpiger) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: I think the user should not set it to a negative number initiatively, but it is better if we can cover these situations~

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala [hunk 2]
```diff
@@ -300,7 +300,8 @@ object PartitioningAwareFileIndex extends Logging {
       sparkSession: SparkSession): Seq[(Path, Seq[FileStatus])] = {
 
     // Short-circuits parallel listing when serial listing is likely to be faster.
-    if (paths.size < sparkSession.sessionState.conf.parallelPartitionDiscoveryThreshold) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala]: why not we just make sure `parallelPartitionDiscoveryThreshold` is greater than 0? We can add a condition(via `checkValue`) in `SQLConf.PARALLEL_PARTITION_DISCOVERY_THRESHOLD`

Comment (windpiger) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala]: sorry, I didn't notice there is a `checkValue` func, let me fix it. thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala [hunk 3]
```diff
@@ -300,7 +300,7 @@ object PartitioningAwareFileIndex extends Logging {
...
+    if (paths.size <= sparkSession.sessionState.conf.parallelPartitionDiscoveryThreshold) {
```
Comment (windpiger) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala]: add equal is more clear to understand the conf `parallelPartitionDiscoveryThreshold `

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala [hunk 4]
```diff
@@ -179,6 +180,21 @@ class FileIndexSuite extends SharedSQLContext {
     }
   }
 
+  test("InMemoryFileIndex with empty rootPaths when PARALLEL_PARTITION_DISCOVERY_THRESHOLD" +
+    "is not positive number") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: `is negative`

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -407,6 +407,8 @@ object SQLConf {
         "files with another Spark distributed job. This applies to Parquet, ORC, CSV, JSON and " +
         "LibSVM data sources.")
       .intConf
+      .checkValue(parallel => parallel >= 0, "The maximum number of files allowed for listing " +
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: actually it should be `The maximum number of root paths`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala (hunk 4 above)
Comment (windpiger) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: there is zero value test in it, so I use not positive...

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: `not positive number` -> `a nonpositive number`

**Comment by gatorsmile** [comment]
LGTM except https://github.com/apache/spark/pull/17093#discussion_r103616280

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala [hunk 6]
```diff
@@ -179,6 +180,21 @@ class FileIndexSuite extends SharedSQLContext {
...
+      new InMemoryFileIndex(spark, Seq.empty, Map.empty, None)
+    }
+
+    val e = intercept[IllegalArgumentException] {
+      withSQLConf(SQLConf.PARALLEL_PARTITION_DISCOVERY_THRESHOLD.key -> "-1") {
+        new InMemoryFileIndex(spark, Seq.empty, Map.empty, None)
+      }
+    }.getMessage
+    assert(e.contains("The maximum number of paths allowed for listing files at " +
+    "driver side must not be negative"))
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: Nit: indent. : )

Comment (windpiger) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileIndexSuite.scala]: oh...

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Pull request closed without merging**